### DISCUSSION TITLE
Filtering

### DISCUSSION THREAD

**Discussion by secretgspot** [bryanmylee/svelte-headless-table#79] [General]
How would one go about filtering within context? 

| name | office |
-----------------
| Sally Jones | Department of Staff |

is it possible to set up filter as such that typing Jones or Staff would pull up proper row?
with default settings ones must type Sally J in order to retrieve that row and just typing Jones or Staff will return none.

**Comment by bryanmylee** [comment]
Hi @secretgspot, this seems like an important use case that I've overlooked. I'll work on extending the table filter plugin.

**Comment by bryanmylee** [comment]
I'm thinking about how the implementation would work, and a rough idea comes to mind.

We have a [`fn`](https://svelte-headless-table.bryanmylee.com/docs/plugins/add-table-filter#fn-filtervalue-value-boolean) option that lets you customize how columns are filtered based on the filter value.

You could then have multiple matches based on this:

```ts
...
tableFilter: addTableFilter({
  fn: ({ filterValue, value }) => {
    const tokens = filterValue.split(' ');
    return tokens.some((token) => 
      value.toLowerCase().startsWith(token.toLowerCase()));
  },
}),
...
```

This would return all rows that had any matches in the filter value tokenized by spaces. However, I'm not sure if we should then sort the matches based on number of matching columns.

An example of how this behaves:

<img width="931" alt="Screenshot 2022-10-25 at 8 23 04 PM" src="https://user-images.githubusercontent.com/42545742/197736263-d9e64202-073d-41c0-bd44-c7664b6f951e.png">

**Reply by secretgspot** [reply]
```js
tableFilter: addTableFilter({
	fn: ({ filterValue, value }) => {
		const tokens = filterValue.split(" ");
		return tokens.some((token) => value.toLowerCase().match(`${token.toLowerCase()}.*`));
	},
}),
```
this another way to filter/match

**Comment by blerrgh** [comment]
For another datapoint, I generally expect filtering to find matches anywhere in the cell value, not just starting at the beginning of a word, like this:

```javascript
fn: ({ filterValue, value }) => {
	return value.includes(filterValue);
},
```

Or case insensitive:

```javascript
fn: ({ filterValue, value }) => {
	return value.toLowerCase().includes(filterValue.toLowerCase());
},
```

Incidentally these also solve the original issue in this thread.
